x/concentrated-liquidity/pool_hooks_test.go

```diff
@@ -178,9 +178,136 @@ func (s *KeeperTestSuite) TestCallPoolActionListener() {
+				ActionPrefix: "invalidActionPrefix",
+				ValidActions: types.GetAllActionPrefixes(),
+			},
+		},
+	}
+
+	for name, tc := range testCases {
+		s.Run(name, func() {
+			s.SetupTest()
+			clPool := s.PrepareConcentratedPool()
+
+			// Upload and instantiate wasm code
+			rawCosmwasmAddress, cosmwasmAddressBech32 := s.uploadAndInstantiateContract(hookContractFilePath)
+
+			// Fund the contract with tokens for all action prefixes using a helper
+			for _, actionPrefix := range tc.actionPrefixes {
+				s.FundAcc(rawCosmwasmAddress, sdk.NewCoins(sdk.NewCoin(actionPrefix, sdk.NewInt(10))))
+			}
+
+			// Set the contract for all hooks as defined by tc.actionPrefixes
+			for _, actionPrefix := range tc.actionPrefixes {
+				// We use the bech32 address here since the set function expects it for security reasons
+				err := s.Clk.SetPoolHookContract(s.Ctx, validPoolId, actionPrefix, cosmwasmAddressBech32)
+
+				if tc.expectedSetError != nil {
+					s.Require().Error(err)
+					s.Require().ErrorContains(tc.expectedSetError, err.Error())
+					return
+				}
+
+				s.Require().NoError(err)
+			}
+
+			// --- Execute a series of actions that trigger all supported hooks if set ---
+
+			// Create position to trigger position creation related hooks
+			_, positionId := s.SetupPosition(clPool.GetId(), s.TestAccs[0], DefaultCoins, types.MinInitializedTick, types.MaxTick, true)
+
+			// Withdraw from position to trigger withdrawal hooks
+			_, _, err := s.Clk.WithdrawPosition(s.Ctx, s.TestAccs[0], positionId, sdk.NewDec(100))
+			s.Require().NoError(err)
+
+			// Fund account and then perform SwapExactAmountIn
+			s.FundAcc(rawCosmwasmAddress, sdk.NewCoins(sdk.NewCoin(types.SwapExactAmountInPrefix, sdk.NewInt(10))))
+			_, err = s.Clk.SwapExactAmountIn(s.Ctx, s.TestAccs[0], clPool, sdk.NewCoin(ETH, sdk.NewInt(1)), USDC, sdk.ZeroInt(), DefaultZeroSpreadFactor)
+			s.Require().NoError(err)
+
+			// Fund account and then perform SwapExactAmountOut
+			s.FundAcc(rawCosmwasmAddress, sdk.NewCoins(sdk.NewCoin(types.SwapExactAmountOutPrefix, sdk.NewInt(10))))
+			_, err = s.Clk.SwapExactAmountOut(s.Ctx, s.TestAccs[0], clPool, ETH, sdk.NewInt(100), sdk.NewCoin(USDC, sdk.NewInt(10)), DefaultZeroSpreadFactor)
+			s.Require().NoError(err)
+
+			// Check that each set hook was successfully triggered.
+			// These assertions lean on the test construction defined in the comments above these tests.
+			// In short, each hook trigger is expected to transfer 1 token with denom corresponding to the
+			// action that triggered it.
+			expectedTriggers := sdk.NewCoins()
+			for _, actionPrefix := range tc.actionPrefixes {
+				expectedTriggers = expectedTriggers.Add(sdk.NewCoin(actionPrefix, sdk.NewInt(1)))
+			}
+
+			balances := s.App.BankKeeper.GetAllBalances(s.Ctx, s.TestAccs[0])
+			s.Require().True(expectedTriggers.DenomsSubsetOf(balances), "expected balance to include: %s, actual balances: %s", expectedTriggers, balances)
```

Comment:
Since we are checking that "expected" is a subset of "actual", we might miss a case where an extra unexpected hook is erroneously triggered.

Could you please look into making this stricter?

Reply:
Added assertion that also asserts that the other possible hooks were not triggered: 3eca42f0ad2222347c6700f5810bb734f3c39941